core/device_hooks/dpcpp_hooks.cpp

```diff
@@ -0,0 +1,118 @@
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+#include <memory>
+#include <string>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/version.hpp>
+
+
+namespace gko {
+
+version version_info::get_dpcpp_version() noexcept
+{
+    // We just return the version with a special "not compiled" tag in
+    // placeholder modules.
+    return {GKO_VERSION_STR, "not compiled"};
+}
+
+
+std::shared_ptr<DpcppExecutor> DpcppExecutor::create(
+    int device_id, std::shared_ptr<Executor> master, std::string device_type)
+{
+    return std::shared_ptr<DpcppExecutor>(
+        new DpcppExecutor(device_id, std::move(master), device_type));
+}
+
+
+void OmpExecutor::raw_copy_to(const DpcppExecutor *, size_type num_bytes,
+                              const void *src_ptr, void *dest_ptr) const
+    GKO_NOT_COMPILED(dpcpp);
+
+
+void DpcppExecutor::raw_free(void *ptr) const noexcept
+{
+    // Free must never fail, as it can be called in destructors.
+    // If the nvidia module was not compiled, the library couldn't have
+    // allocated the memory, so there is no need to deallocate it.
+}
+
+
+void *DpcppExecutor::raw_alloc(size_type num_bytes) const
+    GKO_NOT_COMPILED(dpcpp);
+
+
+void DpcppExecutor::raw_copy_to(const OmpExecutor *, size_type num_bytes,
+                                const void *src_ptr, void *dest_ptr) const
+    GKO_NOT_COMPILED(dpcpp);
+
+
+void DpcppExecutor::raw_copy_to(const CudaExecutor *, size_type num_bytes,
+                                const void *src_ptr, void *dest_ptr) const
+    GKO_NOT_COMPILED(dpcpp);
+
+void DpcppExecutor::raw_copy_to(const HipExecutor *, size_type num_bytes,
+                                const void *src_ptr, void *dest_ptr) const
+    GKO_NOT_COMPILED(dpcpp);
+
+void DpcppExecutor::raw_copy_to(const DpcppExecutor *, size_type num_bytes,
+                                const void *src_ptr, void *dest_ptr) const
+    GKO_NOT_COMPILED(dpcpp);
```

Comment:
nit: Inconsistent empty lines. Suggestion:
```suggestion
void DpcppExecutor::raw_copy_to(const OmpExecutor *, size_type num_bytes,
                                const void *src_ptr, void *dest_ptr) const
    GKO_NOT_COMPILED(dpcpp);


void DpcppExecutor::raw_copy_to(const CudaExecutor *, size_type num_bytes,
                                const void *src_ptr, void *dest_ptr) const
    GKO_NOT_COMPILED(dpcpp);


void DpcppExecutor::raw_copy_to(const HipExecutor *, size_type num_bytes,
                                const void *src_ptr, void *dest_ptr) const
    GKO_NOT_COMPILED(dpcpp);


void DpcppExecutor::raw_copy_to(const DpcppExecutor *, size_type num_bytes,
                                const void *src_ptr, void *dest_ptr) const
    GKO_NOT_COMPILED(dpcpp);
```